README
NAME
    XML::Jing - Validate XML files using an RNG schema using the Jing tool

VERSION
    version 0.03

SYNOPSIS
            use XML::Jing;
            my $jing = XML::Jing->new('path/to/rng','use compact RNG');
            my $error = $jing->validate('path/to/xml');
            if(!$error){
                    print 'no errors!';
            }else{
                    print $error;
            }

DESCRIPTION
    This module is a simple interface to Jing which allows checking XML
    files for validity using an RNG file.

METHODS
  "new"
    Arguments: the path to the RNG file to use in validation, and a boolean
    indicating whether or not the given RNG file uses compact syntax (false
    means XML syntax)

    Creates a new instance of "XML::Jing".

  "validate"
    Argument: path to the XML file to validate

    Returns: The first error found in the document, or "undef" if no errors
    were found.

TODO
    Jing has more functionality and options than what I have interfaced with
    here.

    Also, it would be nice to be able to get ALL of the errors in an XML
    file, instead of jut the first one.

SEE ALSO
    Jing homepage: <http://www.thaiopensource.com/relaxng/jing.html>

    Inline::Java was used to interface with Jing: Inline::Java

AUTHOR
    Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Nathan Glenn.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for XML::Jing


0.03      2013-05-21 15:35:57 America/Los_Angeles
	Better error checking during validation
0.02      2013-05-13 10:37:57 America/Los_Angeles
	Don't include a corrupted jar file
	Use Env::Path instead of Config module
0.01      2013-05-04 04:02:18 UTC
	First release.

Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "File::ShareDir::Install" => "0.03",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Validate XML files using an RNG schema using the Jing tool\r",
  "dist_author" => [
    "Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "XML-Jing",
  "dist_version" => "0.03",
  "license" => "perl",
  "module_name" => "XML::Jing",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Env::Path" => 0,
    "File::ShareDir" => 0,
    "Inline" => 0,
    "Inline::Java" => 0,
    "Path::Tiny" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "share_dir" => {
    "dist" => "java"
  },
  "test_requires" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88"
  }
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  my $tr = delete $module_build_args{test_requires};
  my $br = $module_build_args{build_requires};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/XML/Jing.pm
#
# This file is part of XML-Jing
#
# This software is copyright (c) 2013 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package XML::Jing;
# ABSTRACT: Validate XML files using an RNG schema using the Jing tool
use strict;
use warnings;
use Path::Tiny;
use File::ShareDir 'dist_dir';
use Carp;
our $VERSION = '0.03'; # VERSION


#add the Jing jar to the system classpath
BEGIN{
	use Env::Path;
	my $classpath = Env::Path->CLASSPATH;
	$classpath->Append(path(dist_dir('XML-Jing'),'jing.jar'));
}

require Inline;
Inline->import(
	Java => path(dist_dir('XML-Jing'),'RNGValidator.java'),
	STUDY => ['RNGValidator'],
	);
use Inline::Java qw(caught);


sub new {
	my ($class, $rng_path, $compact) = @_;
	unless (-e $rng_path){
		croak "File doesn't exist: $rng_path";
	}
	my $self = bless {}, $class;

	#read in the RNG file, catching any errors
	eval {
		$self->{validator} = XML::Jing::RNGValidator->new("$rng_path", $compact)
	};
	if ($@){
		if (caught("org.xml.sax.SAXParseException")){
			my $error = 'Error reading RNG file: ' . $@->getMessage();
			#undef $@ so that the Inline::Java object is released (in case someone catches the croak)
			undef $@;
			croak $error;
		}else{
			# It wasn't a Java exception after all...
			my $error = $@;
			undef $@;
			croak $error ;
		}
	}
	return $self;
}


sub validate {
	my ($self, $xml_path) = @_;
	unless (-e $xml_path){
		croak "File doesn't exist: $xml_path";
	}

	#validate the file, catching any errors
	my $errors;
	eval {
		$errors = $self->{validator}->validate("$xml_path")
	};
	if($@){
		if (caught("java.io.FileNotFoundException")){
			my $error = 'Error reading file: ' . $@->getMessage();
			#undef $@ so that the Inline::Java object is released (in case someone catches the croak)
			undef $@;
			croak $error;
		}else{
			warn 'croaking!';
			my $error = $@;
			undef $@;
			croak $error;
		}
	}
	return $errors;
}

1;

__END__

=pod

=head1 NAME

XML::Jing - Validate XML files using an RNG schema using the Jing tool

=head1 VERSION

version 0.03

=head1 SYNOPSIS

	use XML::Jing;
	my $jing = XML::Jing->new('path/to/rng','use compact RNG');
	my $error = $jing->validate('path/to/xml');
	if(!$error){
		print 'no errors!';
	}else{
		print $error;
	}

=head1 DESCRIPTION

This module is a simple interface to Jing which allows checking XML files for validity using an RNG file.

=head1 METHODS

=head2 C<new>

Arguments: the path to the RNG file to use in validation, and a boolean indicating whether or not the given
RNG file uses compact syntax (false means XML syntax)

Creates a new instance of C<XML::Jing>.

=head2 C<validate>

Argument: path to the XML file to validate

Returns: The first error found in the document, or C<undef> if no errors were found.

=head1 TODO

Jing has more functionality and options than what I have interfaced with here.

Also, it would be nice to be able to get ALL of the errors in an XML file, instead of jut the first one.

=head1 SEE ALSO

Jing homepage: L<http://www.thaiopensource.com/relaxng/jing.html>

Inline::Java was used to interface with Jing: L<Inline::Java>

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


java/RNGValidator.java

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;

import javax.xml.XMLConstants;
import javax.xml.transform.stream.StreamSource;
import javax.xml.validation.Schema;
import javax.xml.validation.SchemaFactory;
import javax.xml.validation.Validator;

import org.xml.sax.SAXException;

public class RNGValidator
{
	
	Validator validator;
	/**
	 * 
	 * @param schemaLocation Path to RNG file to be used to validate XML files
	 * @param compact True if the given RNG file uses compact syntax, false if XML.
	 * @throws SAXException 
	 */
	public RNGValidator(String schemaLocation, boolean compact) throws SAXException{
		if(compact)
			System.setProperty(SchemaFactory.class.getName() + ":" + XMLConstants.RELAXNG_NS_URI, "com.thaiopensource.relaxng.jaxp.CompactSyntaxSchemaFactory");
		else
			System.setProperty(SchemaFactory.class.getName() + ":" + XMLConstants.RELAXNG_NS_URI, "com.thaiopensource.relaxng.jaxp.XMLSyntaxSchemaFactory");
		
		SchemaFactory factory = SchemaFactory.newInstance(XMLConstants.RELAXNG_NS_URI);
        Schema schema = factory.newSchema(new File(schemaLocation));

        //get a validator from the schema
        validator = schema.newValidator();
	}

	/**
	 * 
	 * @param xmlFileName Name of XML file to validate
	 * @return Error thrown by validator if the file was invalid, or null if the file was valid
	 * @throws IOException 
	 */
	public String validate(String xmlFileName) throws IOException{

            // Check the document
            try
            {
                validator.validate(new StreamSource(new File(xmlFileName)));
            }
            catch (SAXException ex)
            {
                return ex.getMessage();
            }
            return null;
	}
	
    public static void main(String[] args) throws SAXException, IOException
    {
    	RNGValidator validator = new RNGValidator(args[0],false);
    	System.out.println(validator.validate(args[1]));
    }

}

Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;

use File::ShareDir::Install;
install_share dist => "java";


my %WriteMakefileArgs = (
  "ABSTRACT" => "Validate XML files using an RNG schema using the Jing tool\r",
  "AUTHOR" => "Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "File::ShareDir::Install" => "0.03",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "XML-Jing",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "XML::Jing",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Env::Path" => 0,
    "File::ShareDir" => 0,
    "Inline" => 0,
    "Inline::Java" => 0,
    "Path::Tiny" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}


